polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,281 @@
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Polygeist/IR/Ops.h"
+#include "mlir/Dialect/Polygeist/IR/Polygeist.h"
+#include "mlir/Dialect/Polygeist/Utils/TransformUtils.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "function-specialization"
+
+namespace mlir {
+namespace polygeist {
+#define GEN_PASS_DEF_FUNCTIONSPECIALIZATION
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h.inc"
+} // namespace polygeist
+} // namespace mlir
+
+using namespace mlir;
+using AccessorPtrType = VersionConditionBuilder::AccessorPtrType;
+using AccessorPtrPairType = VersionConditionBuilder::AccessorPtrPairType;
+
+static llvm::cl::opt<unsigned> functionSpecializationAccessorLimit(
+    "function-specialization-accessor-limit", llvm::cl::init(5),
+    llvm::cl::desc(
+        "Maximum number of accessors allowed for function specialization"));
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Returns true if \p type is 'memref<?x!sycl.accessor>', and false otherwise.
+static bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  return isa<sycl::AccessorType>(mt.getElementType());
+}
+
+/// Returns true if \p arg is a candidate argument. Currently, all arguments
+/// with valid memref type determined by isValidMemRefType is a candidate
+/// argument, i.e., 'memref<?x!sycl.accessor>`.
+static bool isCandidateArg(Value arg) {
+  if (isValidMemRefType(arg.getType())) {
+    assert(isa<AccessorPtrType>(arg) &&
+           "Expecting valid memref type to be AccessorPtrType");
+    return true;
+  }
+  return false;
+}
+
+/// Populate \p candArgs with candidate arguments from \p call.
+static void
+collectCandidateArguments(CallOpInterface call,
+                          SmallVectorImpl<AccessorPtrType> &candArgs) {
+  assert(candArgs.empty() && "Expecting empty candArgs");
+  for (Value arg : call.getArgOperands())
+    if (isCandidateArg(arg))
+      candArgs.push_back(cast<AccessorPtrType>(arg));
+}
+
+/// Returns the cloned version of \p func.
+static FunctionOpInterface cloneFunction(FunctionOpInterface func) {
+  FunctionOpInterface clonedFunc = func.clone();
+  std::string newFnName = (func.getName() + ".specialized").str();
+#ifndef NDEBUG
+  ModuleOp module = func->getParentOfType<ModuleOp>();
+  module->walk([newFnName](FunctionOpInterface func) {
+    assert(func.getName() != newFnName &&
+           "Expecting new function name to be unique");
+  });
+#endif // NDEBUG
```

Comment:
This may lead to false positives if we have nested modules. I propose:
1. Check as [here](https://github.com/intel/llvm/blob/9ee75f9465688b46847da46b45fafa3cc1d0adb1/polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp#L1513), which follows `SymbolTable` interface definition;
2. Check early, before actually cloning the function

Reply:
done.